### PR TITLE
Exam - Tea Stefanovska

### PR DESCRIPTION
**Pull request by tejchu**
Practical part 1, has some mistakes.
theoretical part will be submitted soon.

### CODE REVIEW

File: sparseMatrix/test/Test/Matrix/Gen.hs [hunk 1]
```diff
@@ -11,8 +11,8 @@ import qualified Hedgehog.Range as Range
 genNat n = Gen.int (Range.constant 0 n)
 genInt n = Gen.int (Range.constantFrom 0 (-n) n)
 
-genMatrix size maxInt
-  | size <= 0 = genCell 
+genMatrix b maxInt
```
Comment (kajigor) [inline sparseMatrix/test/Test/Matrix/Gen.hs:14]: How does renaming an argument fixes anything?

File: sparseMatrix/test/Test/Matrix/Prop.hs [hunk 2]
```diff
@@ -75,4 +75,47 @@ hprop_scalarMultTranspose = property $ do
...
+  m <- forAll $ genMatrix b 10
+  let unit = diagonalM b 1
+  m === unit `mult` m
+
+hprop_rightIdentityMult :: Property
+hprop_rightIdentityMult = property $ do
+  b <- forAll $ genNat 5
+  m <- forAll $ genMatrix b 10
+  let unit = diagonalM b 1
+  m === m `mult` unit
```
Comment (kajigor) [inline sparseMatrix/test/Test/Matrix/Prop.hs:122]: These properties are not new. Only 2 new properties have been implemented

File: sparseMatrix/src/Matrix.hs [hunk 3]
```diff
@@ -22,4 +22,36 @@ class Num a => Matrix matrix a where
...
+  
+  add (Cell b1 v1) (Cell b2 v2) = Cell b1 (v1 + v2)
+  add (Cell b v) (Quad b' nw ne sw se) = add (Quad b (Cell (b-1) v) (Cell (b-1) v) (Cell (b-1) v) (Cell (b-1) v)) (Quad b' nw ne sw se)
+  add quad@(Quad _ _ _ _ _) cell@(Cell _ _) = add cell quad
+  add (Quad b nw1 ne1 sw1 se1) (Quad _ nw2 ne2 sw2 se2) = Quad b (add nw1 nw2) (add ne1 ne2) (add sw1 sw2) (add se1 se2)
+
+  sub m1 m2 = add m1 (neg m2)
+
+  neg (Cell b v) = Cell b (-v)
+  neg (Quad b nw ne sw se) = Quad b (neg nw) (neg ne) (neg sw) (neg se)
```
Comment (kajigor) [inline sparseMatrix/src/Matrix.hs]: Can be implemented using another matrix operation which would avoid duplication

File: sparseMatrix/src/Matrix.hs [hunk 4]
```diff
@@ -22,4 +22,36 @@ class Num a => Matrix matrix a where
...
+      sw = add (mult sw1 nw2) (mult se1 sw2)
+      se = add (mult sw1 ne2) (mult se1 se2)
+      in Quad b1 nw ne sw se
+    | otherwise = error "Matrixmult: Bounds do not match."
+  mult _ _ = error "Matrixmult: Mismatch in QuadTree dimensions."
+  mult (Quad 0 _ _ _ _) (Quad 0 _ _ _ _) = cell 0 0
+  mult _ (Quad 0 _ _ _ _) = cell 0 0
+  mult (Quad 0 _ _ _ _) _ = cell 0 0
+  mult (Cell 0 _) _ = cell 0 0
+  mult _ (Cell 0 _) = cell 0 0
```
Comment (kajigor) [inline sparseMatrix/src/Matrix.hs:50]: Tests for multiplication fail.

**Review by kajigor — CHANGES_REQUESTED** [review]

File: sparseMatrix/src/Matrix.hs (hunk 3 above)
Comment (tejchu) [reply sparseMatrix/src/Matrix.hs]: This has hopefully been fixed

File: sparseMatrix/src/Matrix.hs (hunk 4 above)
Comment (tejchu) [reply sparseMatrix/src/Matrix.hs:50]: It is only one of the unit tests the one with example matrices, I couldn't figure out what goes wrong

Comment (kajigor) [reply sparseMatrix/src/Matrix.hs:50]: Try multiplying the following matrix by itself with a pen on a paper, compare with what your code computes: 

```
1 1 
1 1 
```

File: sparseMatrix/test/Test/Matrix/Prop.hs (hunk 2 above)
Comment (tejchu) [reply sparseMatrix/test/Test/Matrix/Prop.hs:122]: Could you check if the one I added makes sense?

**Comment by kajigor** [comment]
Now it's 8 out 12.

**Pull request closed without merging**